prqlc/prqlc-parser/src/lexer.rs

```diff
@@ -352,6 +368,76 @@ fn quoted_string(escaped: bool) -> impl Parser<char, String, Error = Cheap<char>
     .labelled("string")
 }
 
+fn interpolated_string(
+    quote: &char,
+    escaping: bool,
+) -> impl Parser<char, Vec<InterpolateItem>, Error = Cheap<char>> + '_ {
+    let opening = just(*quote).repeated().at_least(1);
+
+    opening.then_with(move |opening| {
+        if opening.len() % 2 == 0 {
+            // If we have an even number of quotes, it's an empty string.
+            return chumsky::prelude::empty().to(vec![]).boxed();
+        }
+        let delimiter = just(*quote).repeated().exactly(opening.len());
+
+        let string = if escaping {
+            choice((
+                // Convert double braces to single braces
+                just(vec!['{', '{']).to('{'),
+                just(vec!['}', '}']).to('}'),
+                // Don't allow consuming a backslash, or the start of an expr
+                choice((delimiter, just(vec!['\\']), just(vec!['{']))).not(),
+                escaped_character(),
+                // Or escape the quote char or `{` of the current string (though
+                // maybe we don't need to allow the escaping of `{` given we
+                // have `{{`?)
+                just('\\').ignore_then(just(*quote).or(just('{'))),
+            ))
+            .boxed()
+        } else {
+            choice((
+                just(vec!['{', '{']).to('{'),
+                just(vec!['}', '}']).to('}'),
+                choice((delimiter, just(vec!['{']))).not(),
+            ))
+            .boxed()
+        }
+        .repeated()
+        .at_least(1)
+        .collect()
+        .map(InterpolateItem::String);
+
+        choice((interpolated_expr(), string))
+            .repeated()
+            .at_least(1)
+            .then_ignore(delimiter)
+            .boxed()
+    })
+}
+
+fn interpolated_expr() -> impl Parser<char, InterpolateItem, Error = Cheap<char>> {
+    // Don't allow two opening `{{`, since that's converted to a string of `{`
+    let inner = just('{').repeated().at_least(2).not().rewind().ignore_then(
+        // TODO: decide how we want to handle colons in interpolated expressions
+        // We use rewinds to look ahead and ensure we don't have a closing
+        // bracket (or colon), before forwarding that to the lexer.
+        filter(|c| *c != '}' && *c != ':')
```

Comment:
Are you aware that this prevents interpolations from containing `}`?

Following is not supported:
```
f"result: { my_func_that_takes_a_tuple {a=1} }"
```

Reply:
Good point. I was aware of the issue with colons at least — python also has this problem but doesn't use colons so much.

With closing brackets, that's indeed a more restrictive case, and I agree with your broader point about it being confusing if we don't allow that, given how prominently brackets are used.

One difficulty is that at the lexing stage we don't know about matching opening & closing brackets, so we can't do the thing of "it's the end of the interpolated string if it's the _final_ closing bracket".

We could require escape characters, which seems reasonable if a bit ugly.